cpp/src/arrow/compute/api_scalar.h

```diff
@@ -259,6 +259,27 @@ Result<Datum> IsValid(const Datum& values, ExecContext* ctx = NULLPTR);
 ARROW_EXPORT
 Result<Datum> IsNull(const Datum& values, ExecContext* ctx = NULLPTR);
 
+struct ARROW_EXPORT FillNullOptions : public FunctionOptions {
+  explicit FillNullOptions(Datum fill_value) : fill_value(std::move(fill_value)) {}
+
+  Datum fill_value;
+};
```

Comment:
See comments above. I think this should be implemented as a binary kernel since the "fill values" could be provided by an array. This also will allow the execution layer to (in the near future) insert implicit casts where needed